src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -233,6 +206,51 @@ public static GraphItem<RemoteResolveResult> CreateUnresolvedMatch(LibraryRange
                 return null;
             }
 
+            return await FindPackageLibraryMatchAsync(libraryRange, framework, remoteProviders, localProviders, cacheContext, logger, cancellationToken);
+        }
+
+
+        /// <summary>
+        /// Resolves the library from the given sources. Note that it does not download the package.
+        /// </summary>
+        /// <param name="cache">Cache of requests per library</param>
+        /// <param name="libraryRange">The library requested</param>
+        /// <param name="remoteProviders">remote Providers (all sources, including file sources)</param>
+        /// <param name="localProviders">local providers(gpf, fallback folders)</param>
+        /// <param name="cacheContext">source caching context</param>
+        /// <param name="logger">logger</param>
+        /// <param name="cancellationToken">cancellation token</param>
+        /// <returns>The requested range and remote match.</returns>
+        public static Task<Tuple<LibraryRange, RemoteMatch>> FindPackageLibraryMatchCachedAsync(
+            ConcurrentDictionary<LibraryRange, Task<Tuple<LibraryRange, RemoteMatch>>> cache,
+            LibraryRange libraryRange,
+            IEnumerable<IRemoteDependencyProvider> remoteProviders,
+            IEnumerable<IRemoteDependencyProvider> localProviders,
+            SourceCacheContext cacheContext,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            return cache.GetOrAdd(libraryRange, (cacheKey) => ResolvePackageLibraryMatchAsync(libraryRange, remoteProviders, localProviders, cacheContext, logger, cancellationToken));
+        }
+
+        private static async Task<Tuple<LibraryRange, RemoteMatch>> ResolvePackageLibraryMatchAsync(LibraryRange libraryRange, IEnumerable<IRemoteDependencyProvider> remoteProviders, IEnumerable<IRemoteDependencyProvider> localProviders, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
+        {
+            var match = await FindPackageLibraryMatchAsync(libraryRange, NuGetFramework.AnyFramework, remoteProviders, localProviders, cacheContext, logger, cancellationToken);
+            if (match == null)
+            {
+                match = CreateUnresolvedMatch(libraryRange);
+            }
+            return new Tuple<LibraryRange, RemoteMatch>(libraryRange, match);
+        }
+
+        private static async Task<RemoteMatch> FindPackageLibraryMatchAsync(LibraryRange libraryRange, NuGetFramework framework, IEnumerable<IRemoteDependencyProvider> remoteProviders, IEnumerable<IRemoteDependencyProvider> localProviders, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
```

Comment:
this whole code of introducing 2-3 new APIs and different caching is bit confusing to follow through. Can't we just have a single cache of LibraryRange + framework which could work for both packageRef as well as package download?

Also there are still few more public APIs which can be simplified n made private

Reply:
The tricky thing is that the libraryrange, tfm based caching does a lot more than just discover the library given. 
It ends up downloading the package, reading the nuspec and then potentially doing the same thing for all of it's dependencies by walking the tfm dependency graph. 

For PackageDownload, we really only want to figure out the versions of the nupkg that will be download later on.

The caching has to only be project restore operation wide (like it is now), because of the potential different sources for different projects. 

I can do another pass to see which APIs can be made private again.